### ISSUE TITLE
Volume mesh contains non-tetrahedron cells

### ISSUE THREAD

**Issue by hadiiiil** [nschloe/pygalmesh#212]
Hi, I encountered this problem while trying to generate a volume mesh from a 3D array, which the resulted mesh had tetrahedron cells but also non-tetrahedron cells which is not good for the finite element analysis. I have tried to generate a surface mesh first and then use volume-from-surface-mesh but I encountered the same problem again which is non-tetrahedron cells exist.
I tried to use     delaunay_filter = vtk.vtkDelaunay3D()  to remove the non-tetrahedron cells but the result was that the mesh in not uniform or isotropic.

My question is, is there a way to fix this problem? and why does it raise? And if I can't avoid this issue how can I refine the mesh after removing the non-tetrahedron cells.

![Screenshot 2024-03-20 122848](https://github.com/meshpro/pygalmesh/assets/120106439/925007c9-5329-4f8f-a791-fbbd43c63842)

**Comment by grodgers1** [comment]
Hi, I'm not sure if this is your issue, but the volume-from-surface routine produces both a triangular surface mesh and a tetrahedral volume mesh. Maybe your issue is realted to that? You could try to extract only the tetrahedral elements if necessary. I did this for example using pyvista:

```python
import pyvista as pv

mesh_file = 'test_mesh.vtk'
mesh = pv.read(mesh_file)

celltypes = mesh.celltypes
print(f'Mesh cells: {mesh.n_cells}')
print(f'Number triangles: {np.sum(celltypes==5)}') # see https://docs.pyvista.org/api/utilities/_autosummary/pyvista.celltype
print(f'Number tetrahedra: {np.sum(celltypes==10)}')
print(f'Number other: {np.sum((celltypes!=10) & (celltypes!=5))}')

mesh_tri = mesh.extract_cells_by_type(pv.CellType.TRIANGLE) # surface mesh
mesh_tet = mesh.extract_cells_by_type(pv.CellType.TETRA) # tetrahedral volume mesh
```

You can then export the triangle or tetrahedral mesh alone using e.g. pv.save_meshio()

**Comment by hadiiiil** [comment]
Yes, that's the issue. My senior solved it using the same approach but using VIK library instead.
Thank you for your help.
